main.go

```diff
@@ -0,0 +1,191 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/grandcat/zeroconf"
+)
+
+const (
+	cptvFolder       = "/var/spool/cptv/downloaded"
+	avahiServiceType = "_cacophonator-management._tcp"
+	ledTriggerFile   = "/sys/class/leds/led0/trigger"
+)
+
+type device struct {
+	Name string
+	IPv4 string
+	Port int
+}
+
+func (d device) getRecordingsList() []string {
+	resp, err := http.Get(d.getAddr() + "/api/recordings")
+	if err != nil {
+		log.Println(err)
+		return nil
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		log.Println("non 200 response when getting recordings list")
+		return nil
+	}
+	var ids []string
+	decoder := json.NewDecoder(resp.Body)
+	if err := decoder.Decode(&ids); err != nil {
+		log.Println(err)
+		return nil
+	}
+	return ids
+}
+
+func (d device) getRecording(cptvFolder, id string) error {
+	setLedState("blinking")
+	resp, err := http.Get(d.getAddr() + "/api/recording/" + id)
+	if err != nil {
+		log.Println(err)
+		return nil
+	}
+	defer resp.Body.Close()
+
+	out, err := os.Create(path.Join(cptvFolder, d.Name+"_"+id))
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+
+	return d.deleteRecording(id)
+}
+
+func (d device) deleteRecording(id string) error {
+	req, err := http.NewRequest("DELETE", d.getAddr()+"/api/recording/"+id, nil)
+	client := new(http.Client)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return errors.New("non 200 status code")
+	}
+	return nil
+}
+
+func (d device) getRecordings(cptvFolder string) {
+	log.Printf("searching for recordings on '%s'", d.Name)
+	for _, id := range d.getRecordingsList() {
+		log.Printf("getting recording '%s'", id)
+		err := d.getRecording(cptvFolder, id)
+		if err != nil {
+			log.Println(err)
+		}
+	}
+}
+func (d device) getAddr() string {
+	return fmt.Sprintf("http://%s:%d", d.IPv4, d.Port)
+}
+
+var ledStates = map[string]string{
+	"blinking": "timer",
+	"off":      "none",
+	"on":       "default-on",
+}
+
+func main() {
+	err := runMain()
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+}
+
+func runMain() error {
+	log.SetFlags(0) // Removes default timestamp SetFlags
+	os.MkdirAll(cptvFolder, 0755)
+	setLedState("off")
+	defer setLedState("off")
+	for {
+		devices := getDevices()
+		for _, device := range devices {
+			device.getRecordings(cptvFolder)
+		}
+		if len(devices) > 0 {
+			setLedState("on")
+		} else {
+			setLedState("off")
+		}
+	}
```

Comment:
This approach is fine for now but eventually I think we'll want to restructure the code so that discovery is always running, even when downloads for some devices are in progress. That way we can have multiple downloads in progress and downloads can be started for a new device as it gets in range even when another one is being handled.

This can be done later though.

Reply:
Have added issues for this.

Reply:
That's great thanks.